metals/src/main/scala/scala/meta/internal/parsing/ClassFinder.scala

```diff
@@ -82,4 +83,136 @@ class ClassFinder(trees: Trees) {
       }
     }
   }
+
+  def resolveClassName(
+      path: AbsolutePath,
+      pos: l.Position,
+      newExtension: String
+  ): Option[String] = {
+    val checkInner: Boolean = newExtension == "class"
+    resolveClassName(path, pos, newExtension, checkInner)
+  }
+
+  private def resolveClassName(
+      path: AbsolutePath,
+      pos: l.Position,
+      newExtension: String,
+      checkInner: Boolean
+  ): Option[String] =
+    for {
+      tree <- trees.get(path)
+      val metaPos = pos.toMeta(tree.pos.input)
+      val fileWithoutExtension = path.filename.stripSuffix(".scala")
+      val TraverseResult(pkg, outer, dfnOpt) = traverseTree(
+        tree,
+        metaPos,
+        checkInner
+      )
+      dfn <- dfnOpt
+    } yield {
+      val filename = {
+        val basename = dfn match {
+          case Definition.Cls(name) => name
+          case Definition.Dfn => s"$fileWithoutExtension$$package"
+        }
+        s"$outer$basename.$newExtension"
+      }
+      (pkg ::: List(filename)).mkString("/")
+    }
+
+  private def traverseTree(
+      tree: Tree,
+      pos: Position,
+      checkInner: Boolean
+  ): TraverseResult = {
+    def extractPkgName(t: Term): List[String] = {
+      @tailrec
+      def loop(t: Term, acc: List[String]): List[String] =
+        t match {
+          case Term.Name(name) => (name :: acc)
+          case Term.Select(term, name) => loop(term, name.value :: acc)
+          case _ => acc
+        }
+      loop(t, Nil)
+    }
+
+    def findStat(
+        stats: List[Stat],
+        ifSome: TraverseResult,
+        ifNone: TraverseResult
+    ) =
+      stats
+        .find(s => s.pos.start <= pos.start && pos.start <= s.pos.end)
+        .orElse(stats.headOption) match {
+        case Some(value) => loop(value, ifSome)
+        case None => ifNone
+      }
+
+    @tailrec
+    def loop(tree: Tree, acc: TraverseResult): TraverseResult = {
+      tree match {
+        case Source(List(pkg: m.Pkg)) => loop(pkg, acc)
+        case Source(stats) => findStat(stats, acc, acc)
+
+        case Pkg(select, (p: Pkg) :: _) =>
+          val pkg = extractPkgName(select)
+          loop(p, acc.extendPackage(pkg))
```

Comment:
You also need to loop over objects/classes since they can be embedded in one another. We shouldn't have a totally separate logic than `findClass` and seems overall more complex than needed.

Why not just have something along the lines of
```scala
// get possible class names
def resolveClassName(
      path: AbsolutePath,
      pos: l.Position,
      newExtension: String
): List[String] = {
  findClass(path, pos) match {
    case Some(cls) => Some(cls)
    case None => findSingleClassInFile().orElse { if (isScala3) Some(s"$filename$$package.$extension") else None } 
  } 
}
```

`findSingleClassInFile` would return Some if we have only one class in file.

Reply:
Or does Tasty not produce the same structure of files as class files?